finetune/nn/target_blocks.py

```diff
@@ -30,6 +31,62 @@ def perceptron(x, ny, config, w_init=None, b_init=None):
         return tf.matmul(x, w) + b
 
 
+def gather_indexes(sequence_tensor, positions):
+    """Gathers the vectors at the specific positions over a minibatch."""
+    sequence_shape = shape_list(sequence_tensor)
+    batch_size = sequence_shape[0]
+    seq_length = sequence_shape[1]
+    width = sequence_shape[2]
+
+    flat_offsets = tf.reshape(
+        tf.range(0, batch_size, dtype=tf.int32) * seq_length, [-1, 1]
+    )
+    flat_positions = tf.reshape(positions + flat_offsets, [-1])
+    flat_sequence_tensor = tf.reshape(sequence_tensor, [batch_size * seq_length, width])
+    output_tensor = tf.gather(flat_sequence_tensor, flat_positions)
+    return output_tensor
+
+
+def masked_language_model(*, X, M, mlm_weights, mlm_positions, mlm_ids, embed_weights, hidden, config, reuse=None):
+    X = merge_leading_dims(X, 3)
+    M = merge_leading_dims(M, 2)
+    hidden = merge_leading_dims(hidden, 3)
+    batch, seq, _ = shape_list(X)
+    with tf.variable_scope('model/masked-language-model'):
+        gathered_hidden = gather_indexes(hidden, mlm_positions)
+        final_proj = tf.layers.dense(
+            gathered_hidden,
+            units=config.n_embed,
+            activation=act_fns[config.act_fn],
+            kernel_initializer=tf.random_normal_initializer(stddev=config.weight_stddev),
+            name='dense'
+        )
+        normed_proj = norm(final_proj, 'LayerNorm')
```

Comment:
If we have the choice here I think its best to use tensorflow's built in layer norm like bert does.

Reply:
Yeah that would be fine.  Should be equivalent though right?